File: Kotlin/HolidayValidator.kt
package holidaychecker

import java.time.LocalDate

/**
 * HolidayValidator
 *
 * This base class is used to determine if a given date is a holiday.
 * It supports fixed holidays, dynamic Islamic holidays, weekends, and custom holidays.
 */
open class HolidayValidator(private val includeSaturday: Boolean) {
    companion object {
        val fixedHolidays = listOf(
            "01-01", // Yeni Yıl
            "04-23", // Ulusal Egemenlik ve Çocuk Bayramı
            "05-01", // İşçi Bayramı
            "05-19", // Atatürk'ü Anma, Gençlik ve Spor Bayramı
            "07-20", // Barış ve Özgürlük Bayramı
            "08-30", // Zafer Bayramı
            "10-29"  // Cumhuriyet Bayramı
        )

        val baseDates = listOf(
            Holiday("ramazanBayrami", LocalDate.of(2024, 4, 10), 3),
            Holiday("kurbanBayrami", LocalDate.of(2024, 6, 28), 4)
        )

        const val islamicYearDays = 354.36667
        const val baseYear = 2024
    }

    private val customHolidays = mutableListOf<CustomHoliday>()

    /**
     * Check if the given date is a holiday
     *
     * @param date The date to check
     * @return True if the date is a holiday, false otherwise
     */
    fun isHoliday(date: LocalDate): Boolean {
        return isWeekend(date) || isFixedHoliday(date) || isIslamicHoliday(date) || isCustomHoliday(date)
    }

    /**
     * Add a custom holiday
     *
     * @param date The date of the custom holiday
     * @param recurring Whether the holiday recurs every year
     * @param isIslamic Whether the holiday should be considered as an Islamic holiday
     */
    fun addCustomHoliday(date: LocalDate, recurring: Boolean = false, isIslamic: Boolean = false) {
        customHolidays.add(CustomHoliday(date, recurring, isIslamic))
    }

    /**
     * Check if the given date is a weekend
     *
     * @param date The date to check
     * @return True if the date is a weekend, false otherwise
     */
    private fun isWeekend(date: LocalDate): Boolean {
        val dayOfWeek = date.dayOfWeek.value
        return dayOfWeek == 7 || (includeSaturday && dayOfWeek == 6)
    }

    /**
     * Check if the given date is a fixed holiday
     *
     * @param date The date to check
     * @return True if the date is a fixed holiday, false otherwise
     */
    private fun isFixedHoliday(date: LocalDate): Boolean {
        val monthDay = date.toString().substring(5)
        return fixedHolidays.contains(monthDay)
    }

    /**
     * Check if the given date is an Islamic holiday
     *
     * @param date The date to check
     * @return True if the date is an Islamic holiday, false otherwise
     */
    private fun isIslamicHoliday(date: LocalDate): Boolean {
        val islamicHolidays = calculateIslamicHolidays(date.year)
        return islamicHolidays.contains(date)
    }

    /**
     * Check if the given date is a custom holiday
     *
     * @param date The date to check
     * @return True if the date is a custom holiday, false otherwise
     */
    private fun isCustomHoliday(date: LocalDate): Boolean {
        return customHolidays.any { holiday ->
            if (holiday.recurring) {
                date.toString().substring(5) == holiday.date.toString().substring(5)
            } else {
                date == holiday.date
            }
        }
    }

    /**
     * Calculate Islamic holidays for a given year
     *
     * @param year The year to calculate the holidays for
     * @return List of Islamic holiday dates
     */
    private fun calculateIslamicHolidays(year: Int): List<LocalDate> {
        val yearDifference = year - baseYear
        return baseDates.flatMap { baseDate ->
            generateHolidayDates(calculateIslamicDate(baseDate.date, yearDifference), baseDate.days)
        }
    }

    /**
     * Calculate the date of an Islamic holiday
     *
     * @param baseDate The base date of the holiday
     * @param yearDifference The difference in years from the base year
     * @return The calculated date of the holiday
     */
    private fun calculateIslamicDate(baseDate: LocalDate, yearDifference: Int): LocalDate {
        val daysToAdd = (yearDifference * islamicYearDays).toInt()
        return baseDate.plusDays(daysToAdd.toLong())
    }

    /**
     * Generate a list of holiday dates starting from a given date
     *
     * @param startDate The start date of the holiday
     * @param days The number of days the holiday lasts
     * @return List of holiday dates
     */
    private fun generateHolidayDates(startDate: LocalDate, days: Int): List<LocalDate> {
        return (0 until days).map { startDate.plusDays(it.toLong()) }
    }

    /**
     * Holiday class representing a holiday with a name, date, and duration in days
     */
    data class Holiday(val name: String, val date: LocalDate, val days: Int)

    /**
     * CustomHoliday class representing a user-defined holiday with additional options for recurrence and Islamic holiday check
     */
    data class CustomHoliday(val date: LocalDate, val recurring: Boolean, val isIslamic: Boolean)
}

File: C#/HolidayValidator.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace HolidayChecker
{
    /// <summary>
    /// HolidayValidator
    ///
    /// This base class is used to determine if a given date is a holiday.
    /// It supports fixed holidays, dynamic Islamic holidays, weekends, and custom holidays.
    /// </summary>
    public abstract class HolidayValidator
    {
        protected static readonly List<string> FixedHolidays = new List<string>
        {
            "01-01", // Yeni Yıl
            "04-23", // Ulusal Egemenlik ve Çocuk Bayramı
            "05-01", // İşçi Bayramı
            "05-19", // Atatürk'ü Anma, Gençlik ve Spor Bayramı
            "07-20", // Barış ve Özgürlük Bayramı
            "08-30", // Zafer Bayramı
            "10-29"  // Cumhuriyet Bayramı
        };

        protected static readonly List<Holiday> BaseDates = new List<Holiday>
        {
            new Holiday("ramazanBayrami", new DateTime(2024, 4, 10), 3),
            new Holiday("kurbanBayrami", new DateTime(2024, 6, 28), 4)
        };

        private const double IslamicYearDays = 354.36667;
        private const int BaseYear = 2024;

        protected bool IncludeSaturday;
        protected readonly List<CustomHoliday> CustomHolidays = new List<CustomHoliday>();

        /// <summary>
        /// Constructor for the HolidayValidator class
        /// </summary>
        /// <param name="includeSaturday">Boolean flag indicating whether to include Saturdays as holidays</param>
        protected HolidayValidator(bool includeSaturday)
        {
            IncludeSaturday = includeSaturday;
        }

        /// <summary>
        /// Check if the given date is a holiday
        /// </summary>
        /// <param name="date">The date to check</param>
        /// <returns>True if the date is a holiday, false otherwise</returns>
        public bool IsHoliday(DateTime date)
        {
            return IsWeekend(date) || IsFixedHoliday(date) || IsIslamicHoliday(date) || IsCustomHoliday(date);
        }

        /// <summary>
        /// Add a custom holiday
        /// </summary>
        /// <param name="date">The date of the custom holiday</param>
        /// <param name="recurring">Whether the holiday recurs every year</param>
        /// <param name="isIslamic">Whether the holiday should be considered as an Islamic holiday</param>
        public void AddCustomHoliday(DateTime date, bool recurring = false, bool isIslamic = false)
        {
            CustomHolidays.Add(new CustomHoliday(date, recurring, isIslamic));
        }

        /// <summary>
        /// Check if the given date is a weekend
        /// </summary>
        /// <param name="date">The date to check</param>
        /// <returns>True if the date is a weekend, false otherwise</returns>
        private bool IsWeekend(DateTime date)
        {
            var dayOfWeek = (int)date.DayOfWeek;
            return dayOfWeek == 0 || (IncludeSaturday && dayOfWeek == 6);
        }

        /// <summary>
        /// Check if the given date is a fixed holiday
        /// </summary>
        /// <param name="date">The date to check</param>
        /// <returns>True if the date is a fixed holiday, false otherwise</returns>
        protected virtual bool IsFixedHoliday(DateTime date)
        {
            return FixedHolidays.Contains(date.ToString("MM-dd"));
        }

        /// <summary>
        /// Check if the given date is an Islamic holiday
        /// </summary>
        /// <param name="date">The date to check</param>
        /// <returns>True if the date is an Islamic holiday, false otherwise</returns>
        private bool IsIslamicHoliday(DateTime date)
        {
            var islamicHolidays = CalculateIslamicHolidays(date.Year);
            return islamicHolidays.Contains(date);
        }

        /// <summary>
        /// Check if the given date is a custom holiday
        /// </summary>
        /// <param name="date">The date to check</param>
        /// <returns>True if the date is a custom holiday, false otherwise</returns>
        private bool IsCustomHoliday(DateTime date)
        {
            return CustomHolidays.Any(holiday =>
            {
                if (holiday.Recurring)
                {
                    return date.ToString("MM-dd") == holiday.Date.ToString("MM-dd");
                }
                return date.Date == holiday.Date.Date;
            });
        }

        /// <summary>
        /// Calculate Islamic holidays for a given year
        /// </summary>
        /// <param name="year">The year to calculate the holidays for</param>
        /// <returns>A list of Islamic holidays</returns>
        private List<DateTime> CalculateIslamicHolidays(int year)
        {
            var yearDifference = year - BaseYear;
            var holidays = new List<DateTime>();

            foreach (var baseDate in BaseDates)
            {
                var startDate = CalculateIslamicDate(baseDate.Date, yearDifference);
                holidays.AddRange(GenerateHolidayDates(startDate, baseDate.Days));
            }

            return holidays;
        }

        /// <summary>
        /// Calculate the date of an Islamic holiday
        /// </summary>
        /// <param name="baseDate">The base date of the holiday</param>
        /// <param name="yearDifference">The difference in years from the base year</param>
        /// <returns>The calculated date of the holiday</returns>
        private DateTime CalculateIslamicDate(DateTime baseDate, int yearDifference)
        {
            return baseDate.AddDays(Math.Round(yearDifference * IslamicYearDays));
        }

        /// <summary>
        /// Generate a list of holiday dates
        /// </summary>
        /// <param name="startDate">The start date of the holiday</param>
        /// <param name="days">The number of days the holiday lasts</param>
        /// <returns>A list of holiday dates</returns>
        private IEnumerable<DateTime> GenerateHolidayDates(DateTime startDate, int days)
        {
            return Enumerable.Range(0, days).Select(i => startDate.AddDays(i));
        }

        /// <summary>
        /// Holiday class representing a holiday with a name, date, and duration
        /// </summary>
        protected class Holiday
        {
            public string Name { get; }
            public DateTime Date { get; }
            public int Days { get; }

            /// <summary>
            /// Constructor for the Holiday class
            /// </summary>
            /// <param name="name">The name of the holiday</param>
            /// <param name="date">The date of the holiday</param>
            /// <param name="days">The number of days the holiday lasts</param>
            public Holiday(string name, DateTime date, int days)
            {
                Name = name;
                Date = date;
                Days = days;
            }
        }

        /// <summary>
        /// CustomHoliday class representing a custom holiday with a date, recurrence flag, and Islamic flag
        /// </summary>
        protected class CustomHoliday
        {
            public DateTime Date { get; }
            public bool Recurring { get; }
            public bool IsIslamic { get; }

            /// <summary>
            /// Constructor for the CustomHoliday class
            /// </summary>
            /// <param name="date">The date of the custom holiday</param>
            /// <param name="recurring">Whether the holiday recurs every year</param>
            /// <param name="isIslamic">Whether the holiday is considered an Islamic holiday</param>
            public CustomHoliday(DateTime date, bool recurring, bool isIslamic)
            {
                Date = date;
                Recurring = recurring;
                IsIslamic = isIslamic;
            }
        }
    }
}
